src/lib.rs

```diff
@@ -155,15 +158,27 @@ pub fn run(mut vm: VMState) -> (U256, VMState) {
                 // TODO: This is not how return works. Fix when we have calls between contracts
                 // hooked up.
                 // This is only to keep the context for tests
-                Variant::Ret(_) => {
-                    if vm.running_contexts.len() > 1
-                        || !vm.current_context().near_call_frames.is_empty()
-                    {
-                        vm.pop_frame()
-                    } else {
-                        break;
+                Variant::Ret(ret_variant) => match ret_variant {
+                    RetOpcode::Ok => {
+                        let should_break = _ok(&mut vm, &opcode);
+                        if should_break {
+                            break;
+                        }
                     }
-                }
+                    RetOpcode::Revert => {
+                        let should_break = _revert(&mut vm, &opcode);
+                        if should_break {
+                            panic!("Contract Reverted");
+                        }
+                    }
+                    RetOpcode::Panic => {
+                        let should_break = _panic(&mut vm, &opcode);
+                        if should_break {
```

Comment:
Same here, also, returning an Err, let's use properly test and report which Error the VM returns. This would also make the run function have to return a Result type, extending this PR a little bit, so let's discuss it a bit, but I think it's time we start returning Errs instead of panics. Starting from this PR.

Reply:
Same as above